Source/Core/Core/IOS/MIOS.cpp

```diff
@@ -0,0 +1,112 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include <cstring>
+#include <vector>
+
+#include "Common/CommonFuncs.h"
+#include "Common/Logging/Log.h"
+#include "Core/Boot/ElfReader.h"
+#include "Core/ConfigManager.h"
+#include "Core/DSPEmulator.h"
+#include "Core/HW/DSP.h"
+#include "Core/HW/Memmap.h"
+#include "Core/HW/SystemTimers.h"
+#include "Core/IOS/MIOS.h"
+#include "Core/PowerPC/PowerPC.h"
+#include "DiscIO/NANDContentLoader.h"
+
+namespace IOS
+{
+namespace HLE
+{
+namespace MIOS
+{
+constexpr u64 MIOS_TITLE_ID = 0x0000000100000101;
+
+static std::vector<u8> GetMIOSBinary()
+{
+  const auto& loader =
+      DiscIO::CNANDContentManager::Access().GetNANDLoader(MIOS_TITLE_ID, Common::FROM_SESSION_ROOT);
+  if (!loader.IsValid())
+    return {};
+
+  const auto* content = loader.GetContentByIndex(loader.GetBootIndex());
+  if (!content)
+    return {};
+
+  return content->m_Data->Get();
+}
+
+static std::vector<u8> GetELF(const std::vector<u8>& mios)
+{
+  ARMBinaryHeader header;
+  if (mios.size() <= sizeof(header))
+    return {};
+
+  std::memcpy(&header, mios.data(), sizeof(header));
+
+  header.header_size = Common::swap32(header.header_size);
+  header.elf_offset = Common::swap32(header.elf_offset);
+  header.elf_size = Common::swap32(header.elf_size);
+
+  if (header.header_size != sizeof(header) || mios.size() <= header.elf_offset + header.elf_size)
+    return {};
```

Comment:
All of this behavior seems like behavior that could be handled by the `ARMBinaryHeader` struct via a constructor and an `IsValid()` member function.